core/src/main/scala/com/yahoo/maha/core/fact/Fact.scala

```diff
@@ -1838,22 +1843,32 @@ case class PublicFactTable private[fact](name: String
       .mapValues(_.map(tpl => tpl._2)
       .to[SortedSet])*/
 
-  private[this] val secondaryDimFactMap: Map[SortedSet[String], SortedSet[Fact]] =
+/*
+  private[this] val secondaryDimFactMap: Map[SortedSet[String], SortedSet[String]] =
     if (this.parentFactTable.isDefined) parentFactTable.get.getSecondaryDimFactMap
     else
-    facts
+*/
+   if (getPowerSetStorage.isEmpty) {
+     facts
       .values
-      .map(f => (f.dimCols.filter(_.annotations.exists(_.isInstanceOf[ForeignKey])).map(col => col.name), f))
+      .map(f => (f.dimCols.filter(_.annotations.exists(_.isInstanceOf[ForeignKey])).map(col => col.name), f.name))
       .par
       .flatMap(tpl => utils.power(tpl._1, tpl._1.size).map(s => (s.to[SortedSet], tpl._2)))
       .toIndexedSeq
       .groupBy(_._1)
       .mapValues(_.map(tpl => tpl._2)
-      .to[SortedSet])
+      .to[SortedSet]).foreach {
```

Comment:
This may force a materialization, try using a view and do all the transformations, and if you are storing underneath then why even sort it ?